README
NAME

    Data::Object - Object Orientation for Perl 5

VERSION

    version 0.59

SYNOPSIS

        use Data::Object;
    
        # returns a code object
        my $object = Data::Object->new(sub{ join ' ', @_ });
    
        # returns true
        $object->isa('Data::Object::Code');
    
        # returns a string object
        my $string = $code->call('Hello', 'World');
    
        # returns a new string object
        $string = $string->split('')->reverse->join('')->uppercase;
    
        # returns a number object (returns true) and outputs "DLROW OLLEH"
        my $result = $string->say;
    
        # returns true
        $result->isa('Data::Object::Number');

DESCRIPTION

    Data::Object is a framework for writing structured and highly
    object-oriented Perl 5 software programs. Additionally, this
    distribution provides classes which wrap Perl 5 native data types and
    provides methods for operating on the data.

EXPORTS

 all

        use Data::Object qw(:all);

    The all export tag will export all exportable functions.

 core

        use Data::Object qw(:core);

    The core export tag will export the exportable functions const, deduce,
    deduce_deep, detract, detract_deep, immutable, load, prototype, reify,
    and throw exclusively.

 data

        use Data::Object qw(:data);

    The data export tag will export all exportable functions whose names
    are prefixed with the word "data".

 type

        use Data::Object qw(:type);

    The type export tag will export all exportable functions whose names
    are prefixed with the word "type".

FUNCTIONS

 const

        # given 1.098765;
    
        const VERSION => 1.098765;

    The const function creates a constant function using the name and
    expression supplied to it. A constant function is a function that does
    not accept any arguments and whose result(s) are deterministic.

 data_array

        # given [2..5];
    
        $object = data_array [2..5];
        $object->isa('Data::Object::Array');

    The data_array function returns a Data::Object::Array instance which
    wraps the provided data type and can be used to perform operations on
    the data. The type_array function is an alias to this function.

 data_code

        # given sub { 1 };
    
        $object = data_code sub { 1 };
        $object->isa('Data::Object::Code');

    The data_code function returns a Data::Object::Code instance which
    wraps the provided data type and can be used to perform operations on
    the data. The type_code function is an alias to this function.

 data_float

        # given 5.25;
    
        $object = data_float 5.25;
        $object->isa('Data::Object::Float');

    The data_float function returns a Data::Object::Float instance which
    wraps the provided data type and can be used to perform operations on
    the data. The type_float function is an alias to this function.

 data_hash

        # given {1..4};
    
        $object = data_hash {1..4};
        $object->isa('Data::Object::Hash');

    The data_hash function returns a Data::Object::Hash instance which
    wraps the provided data type and can be used to perform operations on
    the data. The type_hash function is an alias to this function.

 data_integer

        # given -100;
    
        $object = data_integer -100;
        $object->isa('Data::Object::Integer');

    The data_integer function returns a Data::Object::Object instance which
    wraps the provided data type and can be used to perform operations on
    the data. The type_integer function is an alias to this function.

 data_number

        # given 100;
    
        $object = data_number 100;
        $object->isa('Data::Object::Number');

    The data_number function returns a Data::Object::Number instance which
    wraps the provided data type and can be used to perform operations on
    the data. The type_number function is an alias to this function.

 data_regexp

        # given qr/test/;
    
        $object = data_regexp qr/test/;
        $object->isa('Data::Object::Regexp');

    The data_regexp function returns a Data::Object::Regexp instance which
    wraps the provided data type and can be used to perform operations on
    the data. The type_regexp function is an alias to this function.

 data_scalar

        # given \*main;
    
        $object = data_scalar \*main;
        $object->isa('Data::Object::Scalar');

    The data_scalar function returns a Data::Object::Scalar instance which
    wraps the provided data type and can be used to perform operations on
    the data. The type_scalar function is an alias to this function.

 data_string

        # given 'abcdefghi';
    
        $object = data_string 'abcdefghi';
        $object->isa('Data::Object::String');

    The data_string function returns a Data::Object::String instance which
    wraps the provided data type and can be used to perform operations on
    the data. The type_string function is an alias to this function.

 data_undef

        # given undef;
    
        $object = data_undef undef;
        $object->isa('Data::Object::Undef');

    The data_undef function returns a Data::Object::Undef instance which
    wraps the provided data type and can be used to perform operations on
    the data. The type_undef function is an alias to this function.

 data_universal

        # given 0;
    
        $object = data_universal 0;
        $object->isa('Data::Object::Universal');

    The data_universal function returns a Data::Object::Universal instance
    which wraps the provided data type and can be used to perform
    operations on the data. The type_universal function is an alias to this
    function.

 deduce

        # given qr/\w+/;
    
        $object = deduce qr/\w+/;
        $object->isa('Data::Object::Regexp');

    The deduce function returns a data type object instance based upon the
    deduced type of data provided.

 deduce_deep

        # given {1,2,3,{4,5,6,[-1]}}
    
        $deep = deduce_deep {1,2,3,{4,5,6,[-1]}};
    
        # Data::Object::Hash {
        #     1 => Data::Object::Number ( 2 ),
        #     3 => Data::Object::Hash {
        #          4 => Data::Object::Number ( 5 ),
        #          6 => Data::Object::Array [ Data::Object::Integer ( -1 ) ],
        #     },
        # }

    The deduce_deep function returns a data type object. If the data
    provided is complex, this function traverses the data converting all
    nested data to objects. Note: Blessed objects are not traversed.

 deduce_type

        # given qr/\w+/;
    
        $type = deduce_type qr/\w+/; # REGEXP

    The deduce_type function returns a data type description for the type
    of data provided, represented as a string in capital letters.

 detract

        # given bless({1..4}, 'Data::Object::Hash');
    
        $object = detract $object; # {1..4}

    The detract function returns a value of native type, based upon the
    underlying reference of the data type object provided.

 detract_deep

        # given {1,2,3,{4,5,6,[-1, 99, bless({}), sub { 123 }]}};
    
        my $object = deduce_deep $object;
        my $revert = detract_deep $object; # produces ...
    
        # {
        #     '1' => 2,
        #     '3' => {
        #         '4' => 5,
        #         '6' => [ -1, 99, bless({}, 'main'), sub { ... } ]
        #       }
        # }

    The detract_deep function returns a value of native type. If the data
    provided is complex, this function traverses the data converting all
    nested data type objects into native values using the objects
    underlying reference. Note: Blessed objects are not traversed.

 immutable

        # given [1,2,3];
    
        $object = immutable data_array [1,2,3];
        $object->isa('Data::Object::Array); # via Data::Object::Immutable

    The immutable function makes the data type object provided immutable.
    This function loads Data::Object::Immutable and returns the object
    provided as an argument.

 load

        # given 'List::Util';
    
        $package = load 'List::Util'; # List::Util if loaded

    The load function attempts to dynamically load a module and either dies
    or returns the package name of the loaded module.

 prototype

        # given ('$name' => [is => 'ro']);
    
        my $proto  = data_prototype '$name' => [is => 'ro'];
        my $class  = $proto->create; # via Data::Object::Prototype
        my $object = $class->new(name => '...');

    The prototype function returns a prototype object which can be used to
    generate classes, objects, and derivatives. This function loads
    Data::Object::Prototype and returns an object based on the arguments
    provided.

 reify

        # given [1..9];
    
        $array = reify [1..9]; # Data::Object::Array

    The reify function will determine the type of the value provided and
    return it as a data type object. This method is an alias to the
    deduce_deep function.

 throw

        # given $message;
    
        throw $message; # An exception (...) was thrown in -e at line 1

    The throw function will dynamically load and throw an exception object.
    This function takes all arguments accepted by the
    Data::Object::Exception class.

SEE ALSO

      * Data::Object::Array

      * Data::Object::Class

      * Data::Object::Class::Syntax

      * Data::Object::Code

      * Data::Object::Float

      * Data::Object::Hash

      * Data::Object::Integer

      * Data::Object::Number

      * Data::Object::Role

      * Data::Object::Role::Syntax

      * Data::Object::Regexp

      * Data::Object::Scalar

      * Data::Object::String

      * Data::Object::Undef

      * Data::Object::Universal

      * Data::Object::Autobox

      * Data::Object::Immutable

      * Data::Object::Library

      * Data::Object::Prototype

      * Data::Object::Signatures

AUTHOR

    Al Newkirk <anewkirk@ana.io>

COPYRIGHT AND LICENSE

    This software is copyright (c) 2014 by Al Newkirk.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



CHANGES
Changelog for Data-Object

0.59 2015-11-29T11:46:26
 - Added global types array to library.
 - Added the Proxyable role to support proxy methods and autoloading.
 - Modified POD in SYNOPSIS for existing roles.
 - Added support for alternate Data::Object namespacing.

0.58 2015-10-11T05:48:56
 - v0.58
 - Updated build.
 - Updated SEE ALSO POD.
 - Updated POD for Hash values method.
 - Modify the Hash fold method to support empty hashes.

0.57 2015-10-10T12:07:44
 - v0.57
 - Updated build.
 - Updated codification examples for Array, Hash, and Code packages.

0.56 2015-10-10T11:39:35
 - v0.56
 - Updated build.
 - Updated POD.
 - Updated codification implementation.
 - Added reify function.
 - Refactored codification to to use Sub::Quote.
 - Updated main module synopsis.

0.55 2015-09-27T20:40:42
 - v0.55
 - Updated build.
 - Updated POD for Syntax packages and removed dead code.

0.54 2015-09-27T12:01:24
 - v0.54
 - Updated build.
 - Updated the .travis.yml file to include Perl versions 5.14+.
 - Fixed code object dump method test failure.

0.53 2015-09-26T14:22:18
 - v0.53
 - Updated build.
 - Fixed Type::Tiny::XS type constraint test failure.

0.52 2015-09-26T11:32:55
 - v0.52
 - Updated build.
 - Updated POD for all syntax packages.

0.51 2015-09-26T10:46:49
 - v0.51
 - Updated build.
 - Updated and cleaned-up POD across all files.

0.50 2015-09-26T09:31:12
 - v0.50
 - Updated build.
 - Complete refactoring, massive changes and cleanup.
 - Miscellaneous changes.
 - Imported Data::Object::Signatures into the codebase.
 - Imported Data::Object::Library into the codebase.
 - Updated dist.ini, .gitignore, and POD for Data::Object, etc.
 - Added method and function signatures to type classes.
 - Miscellaneous changes and cleanup.
 - Added support for producing immutable type objects.
 - Make all data type objects throwable.
 - Refactor import statements and foreign function calls.
 - Added Perl 5.14 and Type::Tiny(::Signatues) dependencies.
 - Updated POD for Data::Object.
 - Updated the distribution description.
 - Updated the POD for the new method on each class.
 - Removed the Constructor role.

0.42 2015-09-17T17:15:23
 - v0.42
 - Updated build.
 - Updated gitignore.
 - Miscellaneous changes.
 - Updated the POD for the syntax packages.

0.41 2015-08-21T01:24:42
 - v0.41
 - Updated build.
 - Updated POD for syntax packages.

0.40 2015-08-13T15:35:31
 - v0.40
 - Updated build.
 - Changed behaviour for req and opt functions.
 - Updated POD for Data::Object.
 - Updated POD for const.

0.39 2015-08-13T11:13:32
 - v0.39
 - Updated build.
 - Added req and opt functions to the syntax packages.
 - Removed the alias function.
 - Added tests and POD for the const function.
 - Refactored the constructor for all type classes.

0.38 2015-08-10T16:40:13
 - v0.38
 - Updated build.
 - Updated miscellaneous role descriptions.
 - Updated POD for Role.
 - Updated the POD for the Singleton class.
 - Updated the ABSTRACT for all relevant classes.
 - Update the Data::Object description.
 - Added const function to the core export tag.

0.37 2015-08-09T20:40:09
 - v0.37
 - Updated build.
 - Added the const function, Updated the load and alias functions.
 - Updated POD for teh Exception class.
 - Updated the main SYNOPSIS.
 - Modified constructor for all objects.

0.36 2015-08-08T21:28:07
 - v0.36
 - Updated build.
 - Added SEE ALSO sections to the class and role packages.

0.35 2015-08-08T21:22:17
 - v0.35
 - Updated build.
 - Fixed and documented the defaulter behavior.

0.34 2015-08-08T19:43:13
 - v0.34
 - Updated build.
 - Update the Data::Object description.
 - Updated the alias function POD.

0.33 2015-08-08T17:18:07
 - v0.33
 - Updated build.
 - Fixed attribute altering and defaulter behavior.
 - Updated the SEE ALSO documentation.
 - Updated POD for syntax packages.
 - Added alias function.

0.32 2015-08-07T12:11:12
 - v0.32
 - Updated build.
 - Added "alt", "def", and "defaulter" functions to the syntax packages.

0.31 2015-08-05T04:26:12
 - v0.31
 - Updated build.
 - Fixed and simplified all class/object constructors.

0.30 2015-08-04T04:04:12
 - v0.30
 - Updated build.
 - Fixed documentation for Class, Role, and Exception packages.

0.29 2015-08-04T03:44:24
 - v0.29
 - Added documentation for Class, Role, and Exception packages.
 - Updated build.
 - Added tests for the Syntax DSL packages.
 - Updated Syntax package descriptions.

0.28 2015-08-03T20:53:33
 - v0.28
 - Updated build.
 - Generating documentation for the Syntax DSL packages.

0.27 2015-08-03T20:43:08
 - v0.27
 - Updated build.
 - Added documentation for the Syntax DSL packages.

0.26 2015-08-02T15:32:19
 - v0.26
 - Updated build.
 - Fixed DSLs for Class and Role definitions.

0.25 2015-08-02T15:23:32
 - v0.25
 - Updated build.
 - Added DSLs for Class and Role definitions.

0.24 2015-07-31T22:59:15
 - v0.24
 - Updated build.
 - Fixed exception class and tests.

0.23 2015-07-30T19:01:20
 - v0.23
 - Updated CHANGES.
 - Modified singleton configuration.
 - Updated build.
 - Allow lists to the Hash and Array constructor.
 - Refactor the Hash fold method.

0.22 2015-07-23T15:54:15
 - v0.22
 - Updated build.
 - Added fold/unfold functionality, Replaced Carp::confess with throw,
   Updated POD.

0.21 2015-07-15T07:20:29
 - v0.21
 - Updated build.
 - Added throw function, "core" and "all" exports, and tests.
 - Modified exception and singleton classes.
 - Added additional merge method tests.
 - Updated project files.
 - Added generic exception class.
 - v0.20
 - Added singleton pattern class.

0.20 2015-06-30T01:56:42
 - v0.20
 - Added singleton pattern class.

0.19 2015-06-14T05:20:10
 - v0.19
 - Added export tags and constructor.

0.18 2015-05-27T19:30:01
 - v0.18
 - Added Data::Object::Role::Type role.
 - Fixed POD rendering errors.
 - Added missing type_regexp alias.


lib/Data/Object.pm
# ABSTRACT: Object Orientation for Perl 5
package Data::Object;

use strict;
use warnings;

use 5.014;

use Carp;
use Scalar::Util;
use Sub::Quote;

use Exporter qw(import);

our $NAMESPACE = 'Data::Object';

# VERSION

my @CORE = grep !/^(data|type)_/, our @EXPORT_OK = qw(
    codify
    const
    data_array
    data_code
    data_float
    data_hash
    data_integer
    data_number
    data_regexp
    data_scalar
    data_string
    data_undef
    data_universal
    deduce
    deduce_deep
    deduce_type
    detract
    detract_deep
    immutable
    load
    prototype
    reify
    throw
    type_array
    type_code
    type_float
    type_hash
    type_integer
    type_number
    type_regexp
    type_scalar
    type_string
    type_undef
    type_universal
);

our %EXPORT_TAGS = (
    all  => [@EXPORT_OK],
    core => [@CORE],
    data => [grep m/^data_/, @EXPORT_OK],
    type => [grep m/^type_/, @EXPORT_OK],
);

sub new {

    shift and goto &deduce_deep;

}

sub const ($$) {

    my $name = shift;
    my $data = shift;

    my $class = caller(0);
       $class = caller(1) if $NAMESPACE eq $class;

    my $fqsn  = $name =~ /(::|')/ ? $name : "${class}::${name}";

    no strict 'refs';
    no warnings 'redefine';

    *{ $fqsn } = sub () { (ref $data eq "CODE") ? goto &$data : $data };

    return $data;

}

sub codify ($;$) {

    my $code = shift;
    my $refs = shift;

    $code = reify($code);

    if ($code->type eq 'UNDEF') {

        # as you were !!!
        $code = q{ @_ };

    }

    elsif ($code->type eq 'CODE') {

        my $func = $code;

        # perform inception !!!
        $refs->{'$exec'} = \$func;
        $code = q{ goto &{$exec} };

    }

    # (facepalm) purely for backwards compatibility
    my $vars = sprintf 'my ($%s) = @_;', join ',$', 'a'..'z';
    my $body = sprintf '%s do { %s; }', $vars, "$code" // '@_';

    my $func = Sub::Quote::quote_sub($body, ref($refs) ? $refs : {});

    return $func;

}

sub immutable ($) {

    my $class = load("${NAMESPACE}::Immutable");

    unshift @_, $class and goto $class->can('new');

}

sub load ($) {

    my $class = shift;

    my $failed = ! $class || $class !~ /^\w(?:[\w:']*\w)?$/;
    my $loaded;

    my $error = do {
        local $@;
        $loaded = $class->can('new') || eval "require $class; 1";
        $@
    };

    croak "Error attempting to load $class: $error"
        if $error or $failed or not $loaded;

    return $class;

}

sub prototype (@) {

    my $class = load("${NAMESPACE}::Prototype");

    unshift @_, $class and goto $class->can('new');

}

sub reify ($) {

    goto &deduce_deep;

}
sub throw (@) {

    my $class = load("${NAMESPACE}::Exception");

    unshift @_, $class and goto $class->can('throw');

}

sub data_array ($) {

    my $class = load("${NAMESPACE}::Array");

    unshift @_, $class and goto $class->can('new');

}

sub data_code ($) {

    my $class = load("${NAMESPACE}::Code");

    unshift @_, $class and goto $class->can('new');

}

sub data_float ($) {

    my $class = load("${NAMESPACE}::Float");

    unshift @_, $class and goto $class->can('new');

}

sub data_hash ($) {

    my $class = load("${NAMESPACE}::Hash");

    unshift @_, $class and goto $class->can('new');

}

sub data_integer ($) {

    my $class = load("${NAMESPACE}::Integer");

    unshift @_, $class and goto $class->can('new');

}

sub data_number ($) {

    my $class = load("${NAMESPACE}::Number");

    unshift @_, $class and goto $class->can('new');

}

sub data_regexp ($) {

    my $class = load("${NAMESPACE}::Regexp");

    unshift @_, $class and goto $class->can('new');

}

sub data_scalar ($) {

    my $class = load("${NAMESPACE}::Scalar");

    unshift @_, $class and goto $class->can('new');

}

sub data_string ($) {

    my $class = load("${NAMESPACE}::String");

    unshift @_, $class and goto $class->can('new');

}

sub data_undef (;$) {

    my $class = load("${NAMESPACE}::Undef");

    unshift @_, $class and goto $class->can('new');

}

sub data_universal ($) {

    my $class = load("${NAMESPACE}::Universal");

    unshift @_, $class and goto $class->can('new');

}

sub deduce ($) {

    my $data = shift;

    # return undefined
    if (not defined $data) {
        return data_undef $data;
    }

    # handle blessed
    elsif (Scalar::Util::blessed($data)) {
        return data_regexp $data if $data->isa('Regexp');
        return $data;
    }

    # handle defined
    else {

        # handle references
        if (ref $data) {
            return data_array $data if 'ARRAY' eq ref $data;
            return data_hash  $data if 'HASH'  eq ref $data;
            return data_code  $data if 'CODE'  eq ref $data;
        }

        # handle non-references
        else {
            if (Scalar::Util::looks_like_number($data)) {
                return data_float   $data if $data =~ /\./;
                return data_number  $data if $data =~ /^\d+$/;
                return data_integer $data;
            }
            else {
                return data_string $data;
            }
        }

        # handle unhandled
        return data_scalar $data;

    }

    # fallback
    return data_undef $data;

}

sub deduce_deep {

    my @data = @_;

    for my $data (@data) {
        my $type;

        $data = deduce($data);
        $type   = deduce_type($data);

        if ($type and $type eq 'HASH') {
            for my $i (keys %$data) {
                my $val = $data->{$i};
                $data->{$i} = ref($val) ? deduce_deep($val) : deduce($val);
            }
        }

        if ($type and $type eq 'ARRAY') {
            for (my $i = 0; $i < @$data; $i++) {
                my $val = $data->[$i];
                $data->[$i] = ref($val) ? deduce_deep($val) : deduce($val);
            }
        }
    }

    return wantarray ? (@data) : $data[0];

}

sub deduce_type ($) {

    my $data = shift;

    $data = deduce $data;

    return "ARRAY"     if $data->isa("${NAMESPACE}::Array");
    return "HASH"      if $data->isa("${NAMESPACE}::Hash");
    return "CODE"      if $data->isa("${NAMESPACE}::Code");

    return "FLOAT"     if $data->isa("${NAMESPACE}::Float");
    return "NUMBER"    if $data->isa("${NAMESPACE}::Number");
    return "INTEGER"   if $data->isa("${NAMESPACE}::Integer");

    return "STRING"    if $data->isa("${NAMESPACE}::String");
    return "SCALAR"    if $data->isa("${NAMESPACE}::Scalar");
    return "REGEXP"    if $data->isa("${NAMESPACE}::Regexp");

    return "UNDEF"     if $data->isa("${NAMESPACE}::Undef");
    return "UNIVERSAL" if $data->isa("${NAMESPACE}::Universal");

    return undef;

}

sub detract ($) {

    my $data = shift;

    $data = deduce $data;

    my $type = deduce_type $data;

    INSPECT:
    return $data unless $type;

    return [@$data] if $type eq 'ARRAY';
    return {%$data} if $type eq 'HASH';
    return $$data   if $type eq 'REGEXP';
    return $$data   if $type eq 'FLOAT';
    return $$data   if $type eq 'NUMBER';
    return $$data   if $type eq 'INTEGER';
    return $$data   if $type eq 'STRING';
    return undef    if $type eq 'UNDEF';

    if ($type eq 'SCALAR' or $type eq 'UNIVERSAL') {

        $type = Scalar::Util::reftype($data) // '';

        return [@$data] if $type eq 'ARRAY';
        return {%$data} if $type eq 'HASH';
        return $$data   if $type eq 'FLOAT';
        return $$data   if $type eq 'INTEGER';
        return $$data   if $type eq 'NUMBER';
        return $$data   if $type eq 'REGEXP';
        return $$data   if $type eq 'SCALAR';
        return $$data   if $type eq 'STRING';
        return undef    if $type eq 'UNDEF';

        if ($type eq 'REF') {
            $type = deduce_type($data = $$data)
                and goto INSPECT;
        }

    }

    if ($type eq 'CODE') {
        return sub { goto &{$data} };
    }

    return undef;

}

sub detract_deep {

    my @data = @_;

    for my $data (@data) {
        $data = detract($data);

        if ($data and 'HASH' eq ref $data) {
            for my $i (keys %$data) {
                my $val = $data->{$i};
                $data->{$i} = ref($val) ? detract_deep($val) : detract($val);
            }
        }

        if ($data and 'ARRAY' eq ref $data) {
            for (my $i = 0; $i < @$data; $i++) {
                my $val = $data->[$i];
                $data->[$i] = ref($val) ? detract_deep($val) : detract($val);
            }
        }
    }

    return wantarray ? (@data) : $data[0];

}

{

    # aliases
    no warnings 'once';

    *type_array     = *data_array;
    *type_code      = *data_code;
    *type_float     = *data_float;
    *type_hash      = *data_hash;
    *type_integer   = *data_integer;
    *type_number    = *data_number;
    *type_regexp    = *data_regexp;
    *type_scalar    = *data_scalar;
    *type_string    = *data_string;
    *type_undef     = *data_undef;
    *type_universal = *data_universal;

}

1;

=encoding utf8

=head1 SYNOPSIS

    use Data::Object;

    # returns a code object
    my $object = Data::Object->new(sub{ join ' ', @_ });

    # returns true
    $object->isa('Data::Object::Code');

    # returns a string object
    my $string = $code->call('Hello', 'World');

    # returns a new string object
    $string = $string->split('')->reverse->join('')->uppercase;

    # returns a number object (returns true) and outputs "DLROW OLLEH"
    my $result = $string->say;

    # returns true
    $result->isa('Data::Object::Number');

=head1 DESCRIPTION

Data::Object is a framework for writing structured and highly object-oriented
Perl 5 software programs. Additionally, this distribution provides classes
which wrap Perl 5 native data types and provides methods for operating on the
data.

=cut

=export all

    use Data::Object qw(:all);

The all export tag will export all exportable functions.

=cut

=export core

    use Data::Object qw(:core);

The core export tag will export the exportable functions C<const>, C<deduce>,
C<deduce_deep>, C<detract>, C<detract_deep>, C<immutable>, C<load>, C<prototype>,
C<reify>, and C<throw> exclusively.

=cut

=export data

    use Data::Object qw(:data);

The data export tag will export all exportable functions whose names are
prefixed with the word "data".

=cut

=export type

    use Data::Object qw(:type);

The type export tag will export all exportable functions whose names are
prefixed with the word "type".

=cut

=function const

    # given 1.098765;

    const VERSION => 1.098765;

The const function creates a constant function using the name and expression
supplied to it. A constant function is a function that does not accept any
arguments and whose result(s) are deterministic.

=cut

=function data_array

    # given [2..5];

    $object = data_array [2..5];
    $object->isa('Data::Object::Array');

The data_array function returns a L<Data::Object::Array> instance which wraps
the provided data type and can be used to perform operations on the data. The
C<type_array> function is an alias to this function.

=cut

=function data_code

    # given sub { 1 };

    $object = data_code sub { 1 };
    $object->isa('Data::Object::Code');

The data_code function returns a L<Data::Object::Code> instance which wraps the
provided data type and can be used to perform operations on the data. The
C<type_code> function is an alias to this function.

=cut

=function data_float

    # given 5.25;

    $object = data_float 5.25;
    $object->isa('Data::Object::Float');

The data_float function returns a L<Data::Object::Float> instance which wraps
the provided data type and can be used to perform operations on the data. The
C<type_float> function is an alias to this function.

=cut

=function data_hash

    # given {1..4};

    $object = data_hash {1..4};
    $object->isa('Data::Object::Hash');

The data_hash function returns a L<Data::Object::Hash> instance which wraps the
provided data type and can be used to perform operations on the data. The
C<type_hash> function is an alias to this function.

=cut

=function data_integer

    # given -100;

    $object = data_integer -100;
    $object->isa('Data::Object::Integer');

The data_integer function returns a L<Data::Object::Object> instance which wraps
the provided data type and can be used to perform operations on the data. The
C<type_integer> function is an alias to this function.

=cut

=function data_number

    # given 100;

    $object = data_number 100;
    $object->isa('Data::Object::Number');

The data_number function returns a L<Data::Object::Number> instance which wraps
the provided data type and can be used to perform operations on the data. The
C<type_number> function is an alias to this function.

=cut

=function data_regexp

    # given qr/test/;

    $object = data_regexp qr/test/;
    $object->isa('Data::Object::Regexp');

The data_regexp function returns a L<Data::Object::Regexp> instance which wraps
the provided data type and can be used to perform operations on the data. The
C<type_regexp> function is an alias to this function.

=cut

=function data_scalar

    # given \*main;

    $object = data_scalar \*main;
    $object->isa('Data::Object::Scalar');

The data_scalar function returns a L<Data::Object::Scalar> instance which wraps
the provided data type and can be used to perform operations on the data. The
C<type_scalar> function is an alias to this function.

=cut

=function data_string

    # given 'abcdefghi';

    $object = data_string 'abcdefghi';
    $object->isa('Data::Object::String');

The data_string function returns a L<Data::Object::String> instance which wraps
the provided data type and can be used to perform operations on the data. The
C<type_string> function is an alias to this function.

=cut

=function data_undef

    # given undef;

    $object = data_undef undef;
    $object->isa('Data::Object::Undef');

The data_undef function returns a L<Data::Object::Undef> instance which wraps
the provided data type and can be used to perform operations on the data. The
C<type_undef> function is an alias to this function.

=cut

=function data_universal

    # given 0;

    $object = data_universal 0;
    $object->isa('Data::Object::Universal');

The data_universal function returns a L<Data::Object::Universal> instance which
wraps the provided data type and can be used to perform operations on the data.
The C<type_universal> function is an alias to this function.

=cut

=function deduce

    # given qr/\w+/;

    $object = deduce qr/\w+/;
    $object->isa('Data::Object::Regexp');

The deduce function returns a data type object instance based upon the deduced
type of data provided.

=cut

=function deduce_deep

    # given {1,2,3,{4,5,6,[-1]}}

    $deep = deduce_deep {1,2,3,{4,5,6,[-1]}};

    # Data::Object::Hash {
    #     1 => Data::Object::Number ( 2 ),
    #     3 => Data::Object::Hash {
    #          4 => Data::Object::Number ( 5 ),
    #          6 => Data::Object::Array [ Data::Object::Integer ( -1 ) ],
    #     },
    # }

The deduce_deep function returns a data type object. If the data provided is
complex, this function traverses the data converting all nested data to objects.
Note: Blessed objects are not traversed.

=cut

=function deduce_type

    # given qr/\w+/;

    $type = deduce_type qr/\w+/; # REGEXP

The deduce_type function returns a data type description for the type of data
provided, represented as a string in capital letters.

=cut

=function detract

    # given bless({1..4}, 'Data::Object::Hash');

    $object = detract $object; # {1..4}

The detract function returns a value of native type, based upon the underlying
reference of the data type object provided.

=cut

=function detract_deep

    # given {1,2,3,{4,5,6,[-1, 99, bless({}), sub { 123 }]}};

    my $object = deduce_deep $object;
    my $revert = detract_deep $object; # produces ...

    # {
    #     '1' => 2,
    #     '3' => {
    #         '4' => 5,
    #         '6' => [ -1, 99, bless({}, 'main'), sub { ... } ]
    #       }
    # }

The detract_deep function returns a value of native type. If the data provided
is complex, this function traverses the data converting all nested data type
objects into native values using the objects underlying reference. Note:
Blessed objects are not traversed.

=cut

=function immutable

    # given [1,2,3];

    $object = immutable data_array [1,2,3];
    $object->isa('Data::Object::Array); # via Data::Object::Immutable

The immutable function makes the data type object provided immutable. This
function loads L<Data::Object::Immutable> and returns the object provided as an
argument.

=cut

=function load

    # given 'List::Util';

    $package = load 'List::Util'; # List::Util if loaded

The load function attempts to dynamically load a module and either dies or
returns the package name of the loaded module.

=cut

=function prototype

    # given ('$name' => [is => 'ro']);

    my $proto  = data_prototype '$name' => [is => 'ro'];
    my $class  = $proto->create; # via Data::Object::Prototype
    my $object = $class->new(name => '...');

The prototype function returns a prototype object which can be used to
generate classes, objects, and derivatives. This function loads
L<Data::Object::Prototype> and returns an object based on the arguments
provided.

=cut

=function reify

    # given [1..9];

    $array = reify [1..9]; # Data::Object::Array

The reify function will determine the type of the value provided and return it
as a data type object. This method is an alias to the C<deduce_deep> function.

=cut

=function throw

    # given $message;

    throw $message; # An exception (...) was thrown in -e at line 1

The throw function will dynamically load and throw an exception object. This
function takes all arguments accepted by the L<Data::Object::Exception> class.

=cut

=head1 SEE ALSO

=over 4

=item *

L<Data::Object::Array>

=item *

L<Data::Object::Class>

=item *

L<Data::Object::Class::Syntax>

=item *

L<Data::Object::Code>

=item *

L<Data::Object::Float>

=item *

L<Data::Object::Hash>

=item *

L<Data::Object::Integer>

=item *

L<Data::Object::Number>

=item *

L<Data::Object::Role>

=item *

L<Data::Object::Role::Syntax>

=item *

L<Data::Object::Regexp>

=item *

L<Data::Object::Scalar>

=item *

L<Data::Object::String>

=item *

L<Data::Object::Undef>

=item *

L<Data::Object::Universal>

=item *

L<Data::Object::Autobox>

=item *

L<Data::Object::Immutable>

=item *

L<Data::Object::Library>

=item *

L<Data::Object::Prototype>

=item *

L<Data::Object::Signatures>

=back

=cut
